### PR TITLE
A4: Palindrome Checker

### PR DESCRIPTION
**Pull request by DhaaraniMohanasundaram**

### CODE REVIEW

File: Training24/Test.c [hunk 1]
```diff
@@ -1,19 +1,148 @@
...
+int main () {
+   TestPhrasePalindromes ();
+   TestReverseNumbers ();
+
+   char phrase[256];
+   char numberStr[256];
+   long long number;
+   long long reversed;
+   int isPalindromeResult;
+   int isOverflow;
```
Comment (RaaneshKumar) [inline Training24/Test.c]: These functionalities must be done using bool, not using int.

File: Training24/Test.c [hunk 2]
```diff
@@ -1,19 +1,148 @@
...
+#define MAGENTA "\x1b[35m"
+#define CYAN "\x1b[36m"
+#define RESET "\x1b[0m"
+
+int main () {
+   TestPhrasePalindromes ();
+   TestReverseNumbers ();
+
+   char phrase[256];
+   char numberStr[256];
```
Comment (RaaneshKumar) [inline Training24/Test.c]: Initializing variables of same datatype can be done in a single line. Make this change wherever applicable.

File: Training24/Test.c [hunk 3]
```diff
@@ -1,19 +1,148 @@
...
+            }
+            else {
+               isOverflow = (number < INT_MIN || number > INT_MAX);
+               if (isOverflow) {
+                  printf ("Output: %s\n", MAGENTA "Overflow" RESET);
+                  printf ("Result: %s\n", YELLOW "Failed to Process" RESET);
+               }
+               else {
+                  printf ("Output: %lld\n", reversed);
+                  printf ("Result: %s\n", isPalindromeResult ? GREEN "It's a Palindrome" RESET : RED "Not a Palindrome" RESET);
```
Comment (RaaneshKumar) [inline Training24/Test.c]: Try moving all these print statements into a method. Also, explore enums as discussed.

**Review by RaaneshKumar — CHANGES_REQUESTED** [review]

File: Training24/Program.c [hunk 4]
```diff
@@ -1,15 +1,49 @@
...
+// Program has functions that checks if the phrase is a palindrome or not and
+// Reverse the number and then checks for palindrome.
+// --------------------------------------------------------------------------------
+#include <ctype.h>
+#include <stdbool.h>
+#include <limits.h>
 
-const char* GetMessage() {
-return "Hello, World!";
+int CheckPhrasePalindrome (const char* phrase) {
```
Comment (RaaneshKumar) [inline Training24/Program.c]: This must also be a bool returning function.

**Comment by RaaneshKumar** [comment]
image

When you are handling special characters, the input must be handled in a meaningful and a generic way. In this case, there is no differentiation in why the former is a palindrome and the latter is not.

File: Training24/Test.c [hunk 5]
```diff
@@ -1,19 +1,172 @@
...
+      if (number < 0) IntegerTableRow (number, "Invalid(Negative number)", false);
+      else {
+         int reversedNumber;
+         bool overflow = ReverseAndCheckOverflow (number, &reversedNumber);
+         char outputStr[256];
+         if (overflow) IntegerTableRow (number, "Integer Overflow", false);
+         else {
+            sprintf (outputStr, "%d", reversedNumber);
+            bool isPalindrome = (number == reversedNumber);
+            IntegerTableRow (number, outputStr, isPalindrome);
```
Comment (RaaneshKumar) [inline Training24/Test.c]:
```suggestion
            IntegerTableRow (number, outputStr, number == reversedNumber);
```

File: Training24/Test.c [hunk 6]
```diff
@@ -1,19 +1,172 @@
...
+         }
+      }
+   }
+}
+
+void clearInputBuffer () {
```
Comment (RaaneshKumar) [inline Training24/Test.c]: Function name starts with an uppercase letter. Kindly go through the code once before pushing to avoid these basic mistakes.

File: Training24/Test.c [hunk 7]
```diff
@@ -1,19 +1,172 @@
...
+         break;
+      case 3:
+         printf ("Exiting program.\n");
+         return 0;
+      }
+   }
+   return 0;
+}
+
+void PhraseTableHeader () {
```
Comment (RaaneshKumar) [inline Training24/Test.c]: Function name must convey the purpose of the function. It must be PrintTableHeader (). Make this change wherever applicable.
Also, for printing headers one function is enough. No need of separate functions for printing integer headers and phrase headers.

File: Training24/Test.c [hunk 8]
```diff
@@ -1,19 +1,172 @@
...
+      else {
+         int validNumber = (int)inputNumber; // Cast to int after valid input
+         int reversedNumber;
+         bool overflow = ReverseAndCheckOverflow (validNumber, &reversedNumber);
+         if (overflow) printf ("Result: %s\n", YELLOW "Invalid (Out of range)" RESET);
+         else {
+            bool isPalindrome = IsNumPalindrome (validNumber);
+            printf ("Output: %d\nResult: %s\n", reversedNumber,
+               isPalindrome ? GREEN "It's a Palindrome" RESET : RED "Not a Palindrome" RESET);
+            break;
```
Comment (RaaneshKumar) [inline Training24/Test.c]: You can continue even if it is a valid input. Otherwise, the user would have to repeat making choices for each input. Have a separate key to exit the loop and ask the user to choose again if the user wishes to.

**Review by RaaneshKumar — CHANGES_REQUESTED** [review]

File: Training24/Test.c [hunk 9]
```diff
@@ -1,19 +1,210 @@
...
+         break;
+      case 3:
+         printf ("\nExiting program.\n");
+         return 0;
+      }
+   }
+   return 0;
+}
+
+void ClearInputBuffer () {
```
Comment (RaaneshKumar) [inline Training24/Test.c]: This can be written inside the code itself with a comment. No need of a function for single line works which are used only once.

File: Training24/Test.c [hunk 10]
```diff
@@ -1,19 +1,210 @@
...
+         }
+      }
+   }
+}
+
+void TestUserPhrase () {
+   char userPhrase[500];
+   while (1) {
+      printf ("\nEnter a phrase to test (max 499 characters, or type 'r' to return menu): ");
+      if (fgets (userPhrase, sizeof (userPhrase), stdin) == NULL) {
```
Comment (RaaneshKumar) [inline Training24/Test.c:125]: sizeof (userPhrase) is calculated multiple times. We can store it in a variable and use it whenever necessary. Apply this change wherever necessary.

File: Training24/Test.c [hunk 11]
```diff
@@ -1,19 +1,210 @@
...
+      userPhrase[strcspn (userPhrase, "\n")] = 0;
+      if (strcmp (userPhrase, "r") == 0 || strcmp (userPhrase, "R") == 0) return;   // Return to menu command
+      if (strlen (userPhrase) >= sizeof (userPhrase) - 1) {
+         printf (YELLOW "Error: Input exceeds maximum length." RESET "\n");
+         ClearInputBuffer ();
+         continue;
+      }
+      else {
+         bool result = CheckPhrasePalindrome (userPhrase);
+         printf ("Result: %s\n", result ? GREEN "It's a Palindrome" RESET : RED "Not a Palindrome" RESET);
```
Comment (RaaneshKumar) [inline Training24/Test.c]: You can directly check with CheckPhrasePalindrome (). No need of storing it in a variable since it is used only once.

File: Training24/Program.c [hunk 12]
```diff
@@ -1,15 +1,50 @@
...
+      if (onlyAlphaNum[start] != onlyAlphaNum[end]) return false;
+      start++;
+      end--;
+   }
+   return true;
 }
+
+bool ReverseAndCheckOverflow (int number, int* reversedNumber) {
+   // Check if the input number is within valid int range
+   //if (number < INT_MIN || number > INT_MAX) return true;   // Input overflow
```
Comment (RaaneshKumar) [inline Training24/Program.c]: Remove this if this is an unwanted comment.

File: Training24/Header.h [hunk 13]
```diff
@@ -1,17 +1,30 @@
...
+// Palindrome Checker: Program that checks whether the given input(Phrase and integer)
+// is a palindrome and test few cases automatically.
+// --------------------------------------------------------------------------------------
 #pragma once
+#pragma warning (disable:4996)
+
 #define HEADER_H
+#define _CRT_SECURE_NO_WARNINGS
+
+/// <summary>Function to check if a phrase is a palindrome</summary>
```
Comment (RaaneshKumar) [inline Training24/Header.h]: Comments must explain only the functionality of the functions. Need not write 'Function to ...' in every comment. Make this change wherever applicable.

**Review by RaaneshKumar — CHANGES_REQUESTED** [review]

**Comment by RaaneshKumar** [comment]
image

The functionality seems to fail. Can you please check on this?

File: Training24/Program.c [hunk 14]
```diff
@@ -1,15 +1,45 @@
...
-const char* GetMessage() {
-return "Hello, World!";
+bool CheckPhrasePalindrome (const char* phrase) {
+   if (phrase == NULL) return 0;   // Not a palindrome
+   char onlyAlphaNum[256] = "";
+   int j = 0, start = 0, end = j - 1;
+   // Filter out non-alphanumeric characters and convert to lowercase
+   for (int i = 0; phrase[i] != '\0'; i++) if (isalnum (phrase[i])) onlyAlphaNum[j++] = tolower (phrase[i]);
+   onlyAlphaNum[j] = '\0';
+   while (start < end) {
```
Comment (Dakshita-Minocha) [inline Training24/Program.c:20]: The alphanum check and the increment can be performed in the loop. No need to parse the string twice.

File: Training24/Program.c [hunk 15]
```diff
@@ -1,15 +1,45 @@
...
+      if (onlyAlphaNum[start] != onlyAlphaNum[end]) return false;
+      start++;
+      end--;
+   }
+   return true;
 }
+
+bool ReverseAndCheckOverflow (int number, int* reversedNumber) {
```
Comment (Dakshita-Minocha) [inline Training24/Program.c]: Your method is for reversal and that is what its name should indicate. If you are checking for overflow, you can have a separate bit for it or return false (since operation would not have been separately carried out.

File: Training24/Program.c [hunk 16]
```diff
@@ -1,15 +1,45 @@
...
+   *reversedNumber = 0;
+   while (number > 0) {
+      int digit = number % 10;
+      // Check for overflow before reversing the number
+      if (*reversedNumber > (INT_MAX - digit) / 10) return true;   // Indicate overflow during reversal
```
Comment (Dakshita-Minocha) [inline Training24/Program.c]: Makes more sense to return false if operation has not been successfully carried out

File: Training24/Program.c [hunk 17]
```diff
@@ -1,15 +1,45 @@
...
+// Program on branch A4.
+// Checks for phrase palindrome and Reverse the number then checks for palindrome.
+// --------------------------------------------------------------------------------
+#include <ctype.h>
+#include <stdbool.h>
+#include <limits.h>
 
-const char* GetMessage() {
-return "Hello, World!";
+bool CheckPhrasePalindrome (const char* phrase) {
```
Comment (Dakshita-Minocha) [inline Training24/Program.c]: Please look up compile-time polymorphism in C. C (and most other common languages) allow you to declare function of the same name with different parameters. Also called function overloading.

File: Training24/Test.c [hunk 18]
```diff
@@ -1,19 +1,187 @@
...
-printf ("%s\n", message);
-return;
+#define RED "\x1b[31m"
+#define GREEN "\x1b[32m"
+#define YELLOW "\x1b[33m"
+#define CYAN "\x1b[36m"
+#define RESET "\x1b[0m"
+
+/// <summary>Test and print results for palindrome checker with phrases.</summary>
+void TestPhrasePalindromes ();
```
Comment (Dakshita-Minocha) [inline Training24/Test.c]:
```suggestion
void TestPalindromes ();
```

File: Training24/Test.c [hunk 19]
```diff
@@ -1,19 +1,187 @@
...
+
+/// <summary>Test and print results for reversed number palindrome.</summary>
+void TestReverseNumbers ();
```
Comment (Dakshita-Minocha) [inline Training24/Test.c]:
```suggestion
void TestReveral ();
```

File: Training24/Test.c [hunk 20]
```diff
@@ -1,19 +1,187 @@
...
+      case '3':
+         printf ("\nExiting program.\n"); return 0;
+      default:
+         printf ("Invalid choice. Please enter 1, 2, or 3.\n"); break;
+      }
+   }
+   return 0;
+}
+
+enum HeaderType { Phrase, Integer };
```
Comment (Dakshita-Minocha) [inline Training24/Test.c:78]: Nice!

File: Training24/Test.c [hunk 21]
```diff
@@ -1,19 +1,187 @@
...
+void TestPhrasePalindromes () {
+   PrintTableHeader (Phrase);
+   const char* phrases[] = { "tacOcAT", "hello", "D", "Don't nod", "Race!Car", " ", "22 / 02 / 2022",
+      "12345678987654321", "0123456789" },
+      * expectedResults[] = { "Palindrome", "Not a Palindrome", "Palindrome", "Palindrome",
+      "Palindrome", "Palindrome", "Palindrome", "Palindrome", "Not a Palindrome" };
+   for (int i = 0; i < sizeof (phrases) / sizeof (phrases[0]); i++) {
+      int isPhrasePalindrome = CheckPhrasePalindrome (phrases[i]);
+      const char* actualResult = isPhrasePalindrome ? "Palindrome" : "Not a Palindrome",
+         * output = (strcmp (actualResult, expectedResults[i]) == 0) ? GREEN "PASS" RESET : RED "FAIL" RESET;
```
Comment (Dakshita-Minocha) [inline Training24/Test.c]:
```suggestion
         * output = !strcmp (actualResult, expectedResults[i]) ? GREEN "PASS" RESET : RED "FAIL" RESET;
```

File: Training24/Test.c [hunk 22]
```diff
@@ -1,19 +1,187 @@
...
+
+void TestPhrasePalindromes () {
+   PrintTableHeader (Phrase);
+   const char* phrases[] = { "tacOcAT", "hello", "D", "Don't nod", "Race!Car", " ", "22 / 02 / 2022",
+      "12345678987654321", "0123456789" },
+      * expectedResults[] = { "Palindrome", "Not a Palindrome", "Palindrome", "Palindrome",
+      "Palindrome", "Palindrome", "Palindrome", "Palindrome", "Not a Palindrome" };
+   for (int i = 0; i < sizeof (phrases) / sizeof (phrases[0]); i++) {
+      int isPhrasePalindrome = CheckPhrasePalindrome (phrases[i]);
+      const char* actualResult = isPhrasePalindrome ? "Palindrome" : "Not a Palindrome",
```
Comment (Dakshita-Minocha) [inline Training24/Test.c]: Since you are only using the value in one place
```suggestion
      const char* actualResult = CheckPhrasePalindrome (phrases[i])? "Palindrome" : "Not a Palindrome",
```

File: Training24/Test.c [hunk 23]
```diff
@@ -1,19 +1,187 @@
...
+         * output = (strcmp (actualResult, expectedResults[i]) == 0) ? GREEN "PASS" RESET : RED "FAIL" RESET;
+      PhraseTableRow (phrases[i], expectedResults[i], actualResult, output);
```
Comment (Dakshita-Minocha) [inline Training24/Test.c]:
```suggestion
      PrintPhraseTableRow (phrases[i], expectedResults[i], actualResult, output);
```

File: Training24/Test.c [hunk 24]
```diff
@@ -1,19 +1,187 @@
...
+void TestReverseNumbers () {
+   int testCases[] = { 212, -121, 12345, -9876789, 999999999, 2147447412, 2147483647, -2147483648LL }, reversedNumber;
+   const char* expectedOutputs[] = { "Palindrome", "Not a Palindrome", "Not a Palindrome",
+      "Not a Palindrome", "Palindrome", "Palindrome", "Failed to process", "Not a Palindrome" };
+   PrintTableHeader (Integer);
+   for (int i = 0; i < sizeof (testCases) / sizeof (testCases[0]); i++) {
+      int number = testCases[i];
+      char outputStr[256];
+      bool overflow = false;
+      if (number < 0) IntegerTableRow (number, "Invalid(Negative number)", "Not a Palindrome", expectedOutputs[i]);
```
Comment (Dakshita-Minocha) [inline Training24/Test.c]:
```suggestion
      if (number < 0) PrintIntegerTableRow (number, "Invalid(Negative number)", "Not a Palindrome", expectedOutputs[i]);
```

Comment (Dakshita-Minocha) [inline Training24/Test.c]: This logic should not be coded in your test function. Test functions only compare input and output

File: Training24/Test.c [hunk 25]
```diff
@@ -1,19 +1,187 @@
...
+      else {
+         overflow = ReverseAndCheckOverflow (number, &reversedNumber);
+         if (overflow) {
+            strcpy (outputStr, "Integer Overflow");
+            IntegerTableRow (number, outputStr, "Failed to process", expectedOutputs[i]);
```
Comment (Dakshita-Minocha) [inline Training24/Test.c]: assign failed to process as actual output if overflow is detected and print like this.
```suggestion
            IntegerTableRow (number, outputStr, actualOutput, expectedOutputs[i]);
```

File: Training24/Test.c [hunk 26]
```diff
@@ -1,19 +1,187 @@
...
+void TestUserPhrase () {
+   char userPhrase[500];
+   while (1) {
+      printf ("\nEnter a phrase to test (max 499 characters, or type 'r' to return menu): ");
+      if (fgets (userPhrase, sizeof (userPhrase), stdin) == NULL) {
+         printf (YELLOW "Error: Input not received. Exiting function." RESET "\n");
+         return;
+      }
+      userPhrase[strcspn (userPhrase, "\n")] = 0;
+      if (strcmp (userPhrase, "r") == 0 || strcmp (userPhrase, "R") == 0) return;   // Return to menu command
```
Comment (Dakshita-Minocha) [inline Training24/Test.c]:
```suggestion
      if (!strcmp (userPhrase, "r") || !strcmp (userPhrase, "R")) return;   // Return to menu command
```
or 
```suggestion
      if (!(strcmp (userPhrase, "r") && strcmp (userPhrase, "R"))) return;   // Return to menu command
```

File: Training24/Test.c [hunk 27]
```diff
@@ -1,19 +1,187 @@
...
+      if (strlen (userPhrase) >= sizeof (userPhrase) - 1) {
+         printf (YELLOW "Error: Input exceeds maximum length." RESET "\n");
+         while (getchar () != '\n' && getchar () != EOF);
+         continue;
+      }
+      else printf ("Result: %s\n", CheckPhrasePalindrome (userPhrase) ? GREEN "It's a Palindrome" RESET : RED "Not a Palindrome" RESET);
+   }
+}
+
+void TestUserInteger () {
```
Comment (Dakshita-Minocha) [inline Training24/Test.c]: You don't test user input (there is no point). The point of testing is to test your logic (for yourself alone) so you know that the code you're sending to the user has no bugs.

Comment (Dakshita-Minocha) [inline Training24/Test.c]:
```suggestion
void GetUserInteger () {
```

**Review by Dakshita-Minocha — CHANGES_REQUESTED** [review]
image
Please make sure to test and format your code before committing.

File: Training24/Program.c [hunk 28]
```diff
@@ -8,18 +8,17 @@
...
-   int j = 0, start = 0, end = j - 1;
-   // Filter out non-alphanumeric characters and convert to lowercase
-   for (int i = 0; phrase[i] != '\0'; i++) if (isalnum (phrase[i])) onlyAlphaNum[j++] = tolower (phrase[i]);
-   onlyAlphaNum[j] = '\0';
+bool IsPhrasePalindrome (const char* phrase) {
+   if (phrase == NULL) return false;  // Not a palindrome
+   size_t start = 0, end = strlen (phrase) - 1;
    while (start < end) {
-      if (onlyAlphaNum[start] != onlyAlphaNum[end]) return false;
+      while (start < end && !isalnum (phrase[start])) start++;
```
Comment (Dakshita-Minocha) [inline Training24/Program.c]: You are now passing it n^3 times.

File: Training24/Test.c [hunk 29]
```diff
@@ -55,19 +55,25 @@ int main () {
...
-         printf (CYAN "\n\t   -------------  TESTING PHRASE PALINDROME  -------------\n" RESET);
-         TestUserPhrase (); break;
-      case '2':
-         printf (CYAN "\n\t   -------------  TESTING INTEGER PALINDROME  -------------\n" RESET);
-         TestUserInteger (); break;
-      case '3':
-         printf ("\nExiting program.\n"); return 0;
-      default:
-         printf ("Invalid choice. Please enter 1, 2, or 3.\n"); break;
+      while (*start && isspace (*start)) start++;
```
Comment (Dakshita-Minocha) [inline Training24/Test.c]: Check for a trim function

File: Training24/Test.c [hunk 30]
```diff
@@ -90,97 +96,116 @@ void PrintTableHeader (enum HeaderType type) {
    }
 }
 
-void PhraseTableRow (const char* input, const char* expectedOutput, const char* actualOutput, const char* result) {
+void PrintPhraseTableRow (const char* input, const char* expectedOutput, const char* actualOutput, const char* result) {
```
Comment (Dakshita-Minocha) [inline Training24/Test.c:96]: There is no difference in your PrintPhraseTableRow() and PrintIntegerTableRow() functions. You can use the same function for both cases.

File: Training24/Test.c [hunk 31]
```diff
@@ -90,97 +96,116 @@ void PrintTableHeader (enum HeaderType type) {
...
+   printf ("\t| %-19s | %-24s | %-17s | %-17s | %-15s |\n", testCase, outputStr, expectedOutput, actualOutput, resultStr);
    printf ("\t+---------------------+--------------------------+-------------------+-------------------+--------+\n");
 }
 
-void TestPhrasePalindromes () {
+void TestPalindromes () {
    PrintTableHeader (Phrase);
    const char* phrases[] = { "tacOcAT", "hello", "D", "Don't nod", "Race!Car", " ", "22 / 02 / 2022",
-      "12345678987654321", "0123456789" },
+                             "12345678987654321", "0123456789" },
```
Comment (Dakshita-Minocha) [inline Training24/Test.c]: Try to test edge cases and invalid cases as well and see if your function fails anywhere.

File: Training24/Test.c [hunk 32]
```diff
@@ -90,97 +96,116 @@ void PrintTableHeader (enum HeaderType type) {
...
    PrintTableHeader (Integer);
    for (int i = 0; i < sizeof (testCases) / sizeof (testCases[0]); i++) {
-      int number = testCases[i];
       char outputStr[256];
+      int number;
       bool overflow = false;
-      if (number < 0) IntegerTableRow (number, "Invalid(Negative number)", "Not a Palindrome", expectedOutputs[i]);
+      char* endPtr;
+      number = strtol (testCases[i], &endPtr, 10);
+      if (*endPtr != '\0') {
```
Comment (Dakshita-Minocha) [inline Training24/Test.c]: This condition will never be true. This check is only required for user input strings (if you are not using fgets)

File: Training24/Test.c [hunk 33]
```diff
@@ -90,97 +96,116 @@ void PrintTableHeader (enum HeaderType type) {
...
+         strcpy (outputStr, "Invalid(Non Integer)");
+         const char* actualOutput = "Not a Palindrome";
+         PrintIntegerTableRow (0, outputStr, expectedOutputs[i], actualOutput, testCases[i]);
+      }
       else {
-         overflow = ReverseAndCheckOverflow (number, &reversedNumber);
-         if (overflow) {
-            strcpy (outputStr, "Integer Overflow");
-            IntegerTableRow (number, outputStr, "Failed to process", expectedOutputs[i]);
+         if (number < 0) {
```
Comment (Dakshita-Minocha) [inline Training24/Test.c]: These checks must be performed inside your program not the test function. The test function must only check whether your program output is the same as what you expect.

File: Training24/Test.c [hunk 34]
```diff
@@ -90,97 +96,116 @@ void PrintTableHeader (enum HeaderType type) {
...
+            PrintIntegerTableRow (number, outputStr, expectedOutputs[i], actualOutput, testCases[i]);
          }
          else {
-            sprintf (outputStr, "%d", reversedNumber);
-            const char* actualOutput = IsNumPalindrome (number) ? "Palindrome" : "Not a Palindrome";
-            IntegerTableRow (number, outputStr, actualOutput, expectedOutputs[i]);
+            overflow = ReverseAndCheckOverflow (number, &number);
+            if (!overflow) {
+               strcpy (outputStr, "Integer Overflow");
+               const char* actualOutput = "Failed to process";
```
Comment (Dakshita-Minocha) [inline Training24/Test.c]: outStr and actualOutput do not need to be 2 separate variables. They perform the same function.

File: Training24/Test.c [hunk 35]
```diff
@@ -90,97 +96,116 @@ void PrintTableHeader (enum HeaderType type) {
...
       userPhrase[strcspn (userPhrase, "\n")] = 0;
-      if (strcmp (userPhrase, "r") == 0 || strcmp (userPhrase, "R") == 0) return;   // Return to menu command
+      if (!strcmp (userPhrase, "r") || !strcmp (userPhrase, "R")) return;
       if (strlen (userPhrase) >= sizeof (userPhrase) - 1) {
          printf (YELLOW "Error: Input exceeds maximum length." RESET "\n");
          while (getchar () != '\n' && getchar () != EOF);
          continue;
       }
-      else printf ("Result: %s\n", CheckPhrasePalindrome (userPhrase) ? GREEN "It's a Palindrome" RESET : RED "Not a Palindrome" RESET);
+      if (strlen (userPhrase) == 0) {
```
Comment (Dakshita-Minocha) [inline Training24/Test.c]: Always make recommend changes **wherever applicable** **_throughout your program_**.
```suggestion
      if (!strlen (userPhrase)) {
```

File: Training24/Test.c [hunk 36]
```diff
@@ -90,97 +96,116 @@ void PrintTableHeader (enum HeaderType type) {
...
+         printf ("Result: %s\n", GREEN "It's a Palindrome" RESET);
+         continue;
+      }
+      printf ("Result: %s\n", IsPhrasePalindrome (userPhrase)
+         ? GREEN "It's a Palindrome" RESET
+         : RED "Not a Palindrome" RESET);
```
Comment (Dakshita-Minocha) [inline Training24/Test.c]:
```suggestion
      printf ("Result: %s\n", IsPhrasePalindrome (userPhrase) ? GREEN "It's a Palindrome" RESET : RED "Not a Palindrome" RESET);
```

File: Training24/Test.c [hunk 37]
```diff
@@ -90,97 +96,116 @@ void PrintTableHeader (enum HeaderType type) {
...
    }
 }
 
-void TestUserInteger () {
+void GetUserInteger () {
    char userInput[256];
```
Comment (Dakshita-Minocha) [inline Training24/Test.c:146]: You can take out the common code in a new function GetUserInput(), and call that from either function

File: Training24/Test.c [hunk 38]
```diff
@@ -90,97 +96,116 @@ void PrintTableHeader (enum HeaderType type) {
...
          printf ("Result: %s\n", YELLOW "Invalid input (Empty string)" RESET);
          continue;
       }
-      if (strcmp (userInput, "r\n") == 0 || strcmp (userInput, "R\n") == 0) return;   // Return to menu command
+      if (!strcmp (userInput, "r\n") || !strcmp (userInput, "R\n") ) return;   // Return to menu command
       char* endPtr;
-      long inputNumber = strtol (userInput, &endPtr, 10);
-      if (endPtr == userInput || *endPtr != '\n' || inputNumber < 0) {   // Check if input is valid
+      int inputNumber = strtol (userInput, &endPtr, 10);
+      if (endPtr == userInput || *endPtr != '\n' || inputNumber < 0) {
```
Comment (Dakshita-Minocha) [inline Training24/Test.c]: Negative number check must be done by your function, not outside of it. A negative number is not invalid. It is simply not a palindrome.

**Review by Dakshita-Minocha — CHANGES_REQUESTED** [review]

**Review by Dakshita-Minocha — APPROVED** [review]
